lxd/api_cluster.go

```diff
@@ -2233,3 +2241,439 @@ func internalClusterRaftNodeDelete(d *Daemon, r *http.Request) response.Response
+	}
+
+	return response.BadRequest(fmt.Errorf("Unknown action %q", req.Action))
+}
+
+func evacuateClusterMember(d *Daemon, r *http.Request) response.Response {
+	var err error
+	var node db.NodeInfo
+
+	nodeName := mux.Vars(r)["name"]
+
+	// Set node status to EVACUATED
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		// Get the node.
+		node, err = tx.GetNodeByName(nodeName)
+		if err != nil {
+			return errors.Wrap(err, "Failed to get node by name")
+		}
+
+		if node.State == db.ClusterMemberStatePending {
+			return fmt.Errorf("Cannot evacuate or restore a pending node")
+		}
+
+		// Set node status to EVACUATED to prevent instances from being created.
+		err = tx.UpdateNodeStatus(node.ID, db.ClusterMemberStateEvacuated)
+		if err != nil {
+			return errors.Wrap(err, "Failed to update node status")
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	// Do nothing if the node is already evacuated.
+	if node.State == db.ClusterMemberStateEvacuated {
+		return response.SmartError(fmt.Errorf("Node is already evacuated"))
+	}
+
+	reverter := revert.New()
+	defer reverter.Fail()
+
+	// Ensure node is put into its previous state if anything fails.
+	reverter.Add(func() {
+		d.cluster.Transaction(func(tx *db.ClusterTx) error {
+			tx.UpdateNodeStatus(node.ID, db.ClusterMemberStateCreated)
+
+			return nil
+		})
+	})
+
+	// The instances are retrieved in a separate transaction, after the node is in EVACUATED state.
+	var dbInstances []db.Instance
+
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		// If evacuating, consider only the instances on the node which needs to be evacuated.
+		dbInstances, err = tx.GetInstances(db.InstanceFilter{Node: nodeName})
+		if err != nil {
+			return errors.Wrap(err, "Failed to get instances")
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	instances := make([]instance.Instance, len(dbInstances))
+
+	for i, dbInst := range dbInstances {
+		inst, err := instance.LoadByProjectAndName(d.State(), dbInst.Project, dbInst.Name)
+		if err != nil {
+			return response.SmartError(errors.Wrap(err, "Failed to load instance"))
+		}
+
+		instances[i] = inst
+	}
+
+	var targetNodeName string
+	var targetNode db.NodeInfo
+
+	migrateInstances := instance.GetMigratableInstances(d.cluster, instances)
+
+	run := func(op *operations.Operation) error {
+		source, err := cluster.Connect(node.Address, d.endpoints.NetworkCert(), d.serverCert(), r, true)
+		if err != nil {
+			return errors.Wrap(err, "Failed to connect to source")
+		}
+
+		metadata := make(map[string]interface{})
+
+		// Stop all instances if needed
+		for _, inst := range instances {
+			if !inst.IsRunning() {
+				continue
+			}
+
+			metadata["evacuation_progress"] = fmt.Sprintf("Stopping %s", inst.Name())
+			op.UpdateMetadata(metadata)
+
+			err = inst.Stop(false)
+			if err != nil {
+				return errors.Wrapf(err, "Failed to stop instance %q", inst.Name())
+			}
```

Comment:
We need to be much nicer than this :)

We need logic similar to what we do on clean LXD shutdown, so attempt a normal shutdown of the instance with a timeout, then force it if needed.